lib/Optimizer/CodeGen/WireSetsToProfileQIR.cpp

```diff
@@ -419,6 +436,9 @@ void cudaq::opt::addWiresetToProfileQIRPipeline(OpPassManager &pm,
     wopt.convertTo = profile.str();
   pm.addNestedPass<func::FuncOp>(cudaq::opt::createWireSetToProfileQIR(wopt));
   pm.addPass(cudaq::opt::createWireSetToProfileQIRPost());
+  // Perform final cleanup for other dialect conversions (like func.func)
+  pm.addPass(cudaq::opt::createConvertToQIR());
+  cudaq::opt::addQIRProfilePipeline(pm, profile, /*performPrep=*/false);
```

Comment:
My intention was to not run these. I'm not entirely sure why they are needed? This pipeline encodes "these passes must be run together in this order", as a sort of "if you don't do it this way, all bets are off", so this seems like the wrong place from that perspective.

ConvertToQIR's primary purpose is to convert the code to "full QIR" form. And the point of this wireset pass pipeline is to _not_ convert to "full QIR" at all.

There is an independent `cc-to-llvm` pass that does the work of converting the CC dialect to LLVM. That seems like it should be closer to what we really want to use after this pass pipeline.

It's also strange to run the conversion from full QIR to profile _foo_ QIR pass since the input isn't, and will never be, full QIR here.

I assume that there are small side-effects as far as attributes being stuck on the Module, etc. that are the real driver here. It would be nice to know exactly what those are...

Reply:
> I'm not entirely sure why they are needed?

These were needed for to get everything to lower correctly. Rather than copy/paste the items I needed for the small number of test cases I was dealing with, I tried to err on the side of use common code to ensure that it is a bit more future-proof. (At least that was my thinking.)

The first line (`createConvertToQIR`) was added primarily due to needing to convert the `func.func` to `llvm.func`, but there are bunch of other ones that I believe could pop up, such as the conversions happening in and around here:
https://github.com/NVIDIA/cuda-quantum/blob/515d5b1e967ecdc4549774e001323746179122f3/lib/Optimizer/CodeGen/ConvertToQIR.cpp#L177

The second line adds these items: (code link) https://github.com/bmhowe23/cuda-quantum/blob/cf82cb5a868ff8428777372bb4b65caa757ebca6/lib/Optimizer/CodeGen/ConvertToQIRProfile.cpp#L545-L548

See the manually added inline comments for explanations for each one of them.
```cpp
  createConvertToQIRFuncPass(convertTo); // need QIR function attributes
  createQIRToQIRProfilePass(convertTo); // may not be needed but does no harm
  createVerifyQIRProfile(...); // need to verify the generated QIR as the appropriate base profile or adaptive profile
```

There may be other ways to refactor this so that the full QIR path and the wireset QIR path use common code, but I was initially trying to keep the changeset as small as possible.